### PR TITLE
sdk/py: Propagate provider from local and remote component resources

### LINKED ISSUE #12161: Previously working: ResourceOptions.provider not getting passed to ComponentResource resource in python
### What happened?

When using the Component Resources and setting a provider at the parent level they aren't inherited at the component. 

### Expected Behavior

It should pass the provider down. 

### Steps to reproduce

Code will be provided. 

### Output of `pulumi about`

CLI
Version      3.54.0
Go Version   go1.20
Go Compiler  gc

Plugins
NAME        VERSION
aws         5.28.0
awsx        1.0.1
docker      3.6.1
eks         1.0.1
kubernetes  3.23.1
python      unknown
random      4.8.2

Host
OS       ubuntu
Version  20.04
Arch     x86_64

Current Stack: richard-shade-pulumi/aws-eks-cr-poy2/dev

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### LINKED ISSUE #12593: sdk/py: Debug MLC provider breakage from #12292
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

This issue is to track breakage to MLC provider propagation (#12520)
from #12292, which itself was a fix for #12161.

It's necessary to understand the failure before we try to bring back #12292
or an alternative fix for #12161.

From original report:

> For example something like:
>
> ```python
> mlc = MultiLangComponent(
>     "mlc",
>     foo="bar"
>     opts=ResourceOptions(
>         provider=k8s_provider,
>     ),
> )
> ```
>
> It doesn't use the specified provider. It falls back to the default kubeconfig context.

### Affected area/feature

sdk/py

### PR DESCRIPTION
**Pull request by abhinav**
Each commit is individually reviewable.
The first commit just re-applies #12292.

---

This change brings back #12292 (fix for #12161),
which was reverted in #12522 because it broke MLCs (#12520)
and corrects the issue that caused that failure.

- #12161 manifests as local component resources
  not propagating the `provider` option to their children
- #12520 manifests as remote component resources (MLCs)
  not propagating the `provider` option to their children

Roughly, given:

    class MyComponentResource:
      def __init__(self, ...):
        # ...
        self.child = aws.s3.Bucket(..., opts=ResourceOptions(parent=self))

    MyComponentResource(opts=ResourceOptions(provider=my_aws_provider))

Both bugs took the form of the `aws.s3.Bucket` above not getting
`my_aws_provider`.

We previously fixed #12161 by keeping using the provider for the
resource only if its package matched the resource's type; otherwise, we
placed it in the providers bag for the children of that resource.

This had the effect of dropping the provider for component resources
because:

- `_get_providers` returns a provider (None for component resources)
  and a bag of providers for children of the resource.
  It is stored on the resource:

  https://github.com/pulumi/pulumi/blob/f76bd5463b512222b62d9e60191d93ab2a66290e/sdk/python/lib/pulumi/resource.py#L866

- A request is made to `register_resource`, which calls
  `prepare_resource`:

  https://github.com/pulumi/pulumi/blob/f76bd5463b512222b62d9e60191d93ab2a66290e/sdk/python/lib/pulumi/resource.py#L880-L882
  https://github.com/pulumi/pulumi/blob/f76bd5463b512222b62d9e60191d93ab2a66290e/sdk/python/lib/pulumi/runtime/resource.py#L811

- `prepare_resource` re-interprets the `provider` and `providers`
  resource options, based on what's in the **original** `opts` struct
  completely ignoring the second result of `_get_providers`:

  https://github.com/pulumi/pulumi/blob/f76bd5463b512222b62d9e60191d93ab2a66290e/sdk/python/lib/pulumi/runtime/resource.py#L210

So in short, the reason the original change broke was:
It turned `provider=x` into `providers=[x]` when appropriate,
but the consuming code did not consume the transformed option.

To fix this, we put the bag of providers produced by `_get_providers`
back into the `opts` struct so that `prepare_resource` and friends can
use it.

Testing:
An integration test is added that defines an MLC,
and uses it from Pulumi programs in Go, Python, and Node
that instantiate the MLC with no provider, provider option, and providers option,
and verifies that resources were instantiated with the correct providers.

Resolves #12593
Resolves #12161

### CODE REVIEW

**Comment by abhinav** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12639** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12639" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/12639?utm_source=stack-comment).

**Review by Frassle — APPROVED** [review]

File: tests/integration/construct_component_provider_propagation/testcomponent-go/main.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		return nil, err
+	}
+
+	comp.Result = r.Result
+	return &comp, ctx.RegisterResourceOutputs(&comp, pulumi.Map{
+		"result": comp.Result,
+	})
+}
+
+// Rnadom is a custom resource that generates a random string.
```
Comment (justinvp) [inline tests/integration/construct_component_provider_propagation/testcomponent-go/main.go:72]: Nit:

```suggestion
// Random is a custom resource that generates a random string.
```

File: tests/integration/construct_component_provider_propagation/go/main.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	return &provider, err
+}
+
+type FooComponent struct {
+	pulumi.ResourceState
+
+	Result pulumi.StringOutput `pulumi:"result"`
+}
+
+func NewComponent(ctx *pulumi.Context, name string, args *FooComponentArgs, opts ...pulumi.ResourceOption) (*FooComponent, error) {
```
Comment (justinvp) [inline tests/integration/construct_component_provider_propagation/go/main.go:51]: Nit: `NewFooComponent`

FWIW, I would have called it `Component` to match the type token rather than `FooComponent`, but my care level for the naming in these tests is pretty low, so feel free to leave as-is.

File: tests/integration/construct_component_provider_propagation/python/__main__.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+FooComponent("uses_default")
+
+# Should use the provider passed in as an argument.
+FooComponent("uses_provider", opts=pulumi.ResourceOptions(
+    provider=explicit_provider,
+))
+
+# Should use the provider passed in as an argument
+FooComponent("uses_providers", opts=pulumi.ResourceOptions(
+    providers=[explicit_provider],
```
Comment (justinvp) [inline tests/integration/construct_component_provider_propagation/python/__main__.py:45]: Although we don't encourage it, it's possible to specify the providers via a map instead of a list. Wondering if we should have a `uses_providers_map` case for extra coverage.

**Review by justinvp — APPROVED** [review]
LGTM otherwise

File: tests/integration/integration_test.go [hunk 4]
```diff
@@ -736,3 +738,67 @@ description: A Pulumi program testing passphrase config.
 	stdout, _ := e.RunCommand("pulumi", "config", "get", "foo")
 	assert.Contains(t, stdout, "bar")
 }
+
+// Regression test for https://github.com/pulumi/pulumi/issues/12593.
+//
+// Verifies that a "provider" option passed to a remote component
+// is properly propagated to the component's children.
+func TestConstructProviderPropagation(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (abhinav) [reply tests/integration/integration_test.go:746]: Reasonable. Will split.

File: tests/integration/construct_component_provider_propagation/go/main.go (hunk 2 above)
Comment (abhinav) [reply tests/integration/construct_component_provider_propagation/go/main.go:51]: Renamed to Component

File: tests/integration/construct_component_provider_propagation/python/__main__.py (hunk 3 above)
Comment (abhinav) [reply tests/integration/construct_component_provider_propagation/python/__main__.py:45]: Ack. Adding.

**Comment by abhinav** [comment]
bors merge

**Comment by abhinav** [comment]
Oh, oops. The go.mod in my integration test used 1.20. Changing to 1.18.

**Comment by abhinav** [comment]
bors merge

**Comment by abhinav** [comment]
Failing test passed fine locally. Might be flaky. Will retry.

**Comment by abhinav** [comment]
bors retry
